### PR TITLE
Final Integration with ShieldUtils v5

### PR DESCRIPTION
**Pull request by punamdahiya**
Fixes https://github.com/mozilla/shield-cloudstorage/issues/7

### CODE REVIEW

**Comment by punamdahiya** [comment]
Till Shield Utils v5 lands on master, to test PR after merging run:
npm install --save gregglind/shield-studies-addon-utils#cloud-storage
npm run postinstall

File: webextension/background.js [hunk 1]
```diff
@@ -38,37 +38,41 @@ class StudyLifeCycleHandler {
...
+   *
+   * @param {object} studyInfo browser.study.studyInfo object
+   *
+   * @returns {undefined}
    */
-  async enableFeature(studyInfo) {
-    if (studyInfo.timeUntilExpire) {
+  enableFeature(studyInfo) {
+    /* if (studyInfo.timeUntilExpire) {
       browser.alarms.create(studyInfo.timeUntilExpire, () =>
```
Comment (punamdahiya) [inline webextension/background.js]: @gregglind I have commented browser.alarms call as it's throwing undefined error. Is that needed?

File: webextension/background.js [hunk 2]
```diff
@@ -38,37 +38,41 @@ class StudyLifeCycleHandler {
    * call `this.enableFeature` to actually do the feature/experience/ui.
    */
   constructor() {
+    // IMPORTANT:  Listen for onEndStudy first.
     browser.study.onEndStudy.addListener(this.handleStudyEnding);
```
Comment (punamdahiya) [inline webextension/background.js]: @gregglind Tested after pulling latest shieldutils v5, Disable and uninstall of add-on doesn't trigger handleStudyEnding. Is there more to it?

**Comment by punamdahiya** [comment]
@gregglind Submitting for your feedback, I have pulled in latest shield-utils from your repo. Thanks!

**Comment by punamdahiya** [comment]
With https://github.com/mozilla/shield-studies-addon-utils/pull/180 in develop, to test PR
run

npm install --save mozilla/shield-studies-addon-utils#develop
npm run postinstall

**Comment by punamdahiya** [comment]
Tested with latest shield v5 API in develop, handleStudyEnding in background.js is  not getting invoked on uninstall and disable of extension.
enableFeature is getting called successfully on enable and install

**Comment by gregglind** [comment]
(reviewing / debugging)

File: package.json [hunk 3]
```diff
@@ -29,6 +29,7 @@
     "lint": "npm-run-all lint:*",
     "lint:css": "stylelint 'webextension/skin/*.css'",
     "lint:js": "eslint webextension --ext=js,json,jsm",
+    "postinstall": "node_modules/shield-studies-addon-utils/bin/copyStudyUtils.js webextension/privileged",
```
Comment (motin) [inline package.json]: Nit: Can be shortened to `copyStudyUtils webextension/privileged"

Comment (punamdahiya) [reply package.json]: Fix in PR https://github.com/punamdahiya/shield-cloudstorage/pull/1/files

This works for me once I update Path in bash_profile  to $PATH:$(npm bin)

**Comment by punamdahiya** [comment]
@aswan Requesting review of the cloud storage webextension install uninstall flow. Uninstall of addon is handled explicitly inside privileged code here
https://github.com/mozilla/shield-cloudstorage/pull/38/files#diff-549919a14832a5cacd7ae883b94a4db1R28

This is in addition to v5 Shield API life cycle browser.study.onEndStudy listener handled in background.js
https://github.com/punamdahiya/shield-cloudstorage/blob/f8ac4a770fc665051636fc1b0194899a7a568459/webextension/background.js#L130

**Comment by aswan** [comment]
There's a ton of stuff here, I'm having trouble figuring out exactly what you want me to review.  Is it practical to create separate PRs for the parts that you want me to review and then all the rest that I assume somebody else will review?

**Comment by punamdahiya** [comment]
@aswan Updated PR to create separate commits for easy review. PR has three commits
1. Removes generated files
2. Integrate shield utils v5, this commit follows small-study example in shield utils to bring in v5 changes as reviewed with https://github.com/mozilla/shield-studies-addon-utils/tree/develop/examples/small-study 
3.  Install and Uninstall flow commit is handling code that triggers when study is enabled and when study ends. 

Looking for your review and feedback on install and uninstall commit. Thanks!

File: webextension/background.js [hunk 4]
```diff
@@ -38,37 +38,56 @@ class StudyLifeCycleHandler {
...
+     * IMPORTANT:  Listen for `onEndStudy` before calling `browser.study.setup`
+     * because:
+     * - `setup` can end with 'ineligible' due to 'allowEnroll' key in first session.
+     *
+     */
     browser.study.onEndStudy.addListener(this.handleStudyEnding);
     browser.study.onReady.addListener(this.enableFeature);
   }
 
   /**
```
Comment (aswan) [inline webextension/background.js]: If this is something we're going to ship, dead code like this should be removed

File: webextension/background.js [hunk 5]
```diff
@@ -38,37 +38,56 @@ class StudyLifeCycleHandler {
...
+   *
+   * @returns {undefined}
    */
-  async enableFeature(studyInfo) {
+  enableFeature(studyInfo) {
     if (studyInfo.timeUntilExpire) {
-      browser.alarms.create(studyInfo.timeUntilExpire, () =>
-        browser.study.endStudy("expired"),
-      );
+      const alarmName = `${browser.runtime.id}:studyExpiration`;
```
Comment (aswan) [inline webextension/background.js]: this change looks unrelated to the subject of the commit message...

File: webextension/background.js [hunk 6]
```diff
@@ -105,21 +122,31 @@ class StudyLifeCycleHandler {
...
-    switch (ending.reason) {
+    for (const url of ending.urls) {
+      await browser.tabs.create({ url });
+    }
+    switch (ending.endingName) {
+    // could have different actions depending on positive / ending names
     default:
-      this.cleanup();
-      // uninstall the addon?
+      console.log(`the ending: ${ending.endingName}`);
```
Comment (aswan) [inline webextension/background.js]: This should not be in shipped code

File: webextension/privileged/ui/cloudstorage.js [hunk 7]
```diff
@@ -23,25 +23,21 @@ this.cloudstorage = class extends ExtensionAPI {
       context.extension.rootURI.resolve("privileged/ui/CloudDownloadsView.jsm")
     );
 
-    function cleanUpPrefs() {
-      // Ensure cloud storage study prefs are cleared
-      ["lastprompt", "interval.prompt", "storage.key", "api.enabled"].forEach(pref => {
-        Services.prefs.clearUserPref(`cloud.services.${pref}`);
-      });
-    }
+    const cloudStorageEventEmitter = new EventEmitter();
```
Comment (aswan) [inline webextension/privileged/ui/cloudstorage.js]: this appears to be unused

File: webextension/privileged/ui/cloudstorage.js [hunk 8]
```diff
@@ -63,6 +59,15 @@ this.cloudstorage = class extends ExtensionAPI {
           return path;
         },
 
+        cleanUpPrefs() {
```
Comment (aswan) [inline webextension/privileged/ui/cloudstorage.js]: is this used anywhere?

File: webextension/background.js (hunk 5 above)
Comment (punamdahiya) [reply webextension/background.js]: This sets alarm for timeleft until expiration of study. This code is getting invoked on install and when alarm goes off triggers uninstall of add-on.
This code snippet is taken from small study example of shield utils and should be reviewed there as well https://github.com/mozilla/shield-studies-addon-utils/blob/develop/examples/small-study/src/background.js#L80

File: webextension/background.js (hunk 4 above)
Comment (punamdahiya) [reply webextension/background.js]: By dead code you mean not used  cleanup() method, I am leaving it here in case the final shield utils v5 asks for it. I can remove it before shipping

File: webextension/background.js (hunk 6 above)
Comment (punamdahiya) [reply webextension/background.js]: +1, will remove

File: webextension/privileged/ui/cloudstorage.js (hunk 7 above)
Comment (punamdahiya) [reply webextension/privileged/ui/cloudstorage.js]: This is passed to CloudDownloadView.jsm in line 46 where it's used to send telemetry events

File: webextension/privileged/ui/cloudstorage.js (hunk 8 above)
Comment (punamdahiya) [reply webextension/privileged/ui/cloudstorage.js]: As of now it's not and commented in background.js. https://github.com/mozilla/shield-cloudstorage/pull/38/commits/e5f56cd380b632fc9ff9469e23febceede38e89a#diff-7f87af4f00f7747a008141453c7eb047R64

I can remove it before shipping

**Comment by punamdahiya** [comment]
@aswan I have updated PR with your feedback on removing unused code. Please review. Thanks!

File: webextension/privileged/ui/cloudstorage.js (hunk 7 above)
Comment (aswan) [reply webextension/privileged/ui/cloudstorage.js]: Okay but this is in a commit titled "install and uninstall flow" but it looks to not actually be related to the install/uninstall flow.
Who reviewed the first version of this code?

**Comment by aswan** [comment]
@punamdahiya  To be clear, you want me to review the 3rd and 4th commits together right?  I can't find a way in the github UI to view those two commits as a unit, can you squash them together?

File: webextension/privileged/ui/cloudstorage.js (hunk 7 above)
Comment (punamdahiya) [reply webextension/privileged/ui/cloudstorage.js]: You are right it's not related to install/uninstall and showing up because I moved this initialization from line 43 to line 26 in an attempt to organize and keep const declarations together. This code has been reviewed earlier as part of PR  https://github.com/mozilla/shield-cloudstorage/pull/28/ (commit Integrate telemetry probes)

**Comment by punamdahiya** [comment]
@aswan yes, please review install and uninstall flow, I have squashed those into single commit. Thanks!

**Comment by motin** [comment]
> I can't find a way in the github UI to view those two commits as a unit, can you squash them together?

I don't know if this is available via the github UI, but comparisons between arbitrary commits can be found by inserting the commits into an URL of the following structure, separated by `...`: https://github.com/mozilla/shield-cloudstorage/compare/c931ab4177345b8f8949e973f7c71e7c08e3a661...cf22c7fc87a848ec5547a37825a3c9958dc3d6ef

**Comment by punamdahiya** [comment]
@motin Unsigned add-on should work on both beta and release. Can you please check the version, 
WEE works on Firefox 60 and above.

The easiest way to find the branch is look at pref cloud.services.interval.prompt, for Control branch it's value will be 0. An alternate way to is to look at about:telemetry. When add-on is installed, we are logging message 'addon_init' that should show branch.

About variationOverridePreference, my understanding it was functional for v5. Removing this will take away option to force set a variation. Will test again and remove if its redundant.

File: webextension/privileged/study/api.js [hunk 9]
```diff
@@ -1,2055 +0,0 @@
-/* eslint-disable */
```
[the remark this answers is not in the document]
Comment (punamdahiya) [reply webextension/privileged/study/api.js]: Done

File: webextension/privileged/study/schema.json [hunk 10]
```diff
@@ -1,471 +0,0 @@
-[
```
[the remark this answers is not in the document]
Comment (punamdahiya) [reply webextension/privileged/study/schema.json]: Done

File: package.json [hunk 11]
```diff
@@ -14,7 +14,8 @@
     "npm-run-all": "4.1.2",
     "stylelint": "9.2.0",
     "stylelint-config-recommended": "2.1.0",
-    "web-ext": "2.6.0"
+    "shield-studies-addon-utils": "github:mozilla/shield-studies-addon-utils#develop",
```
[the remark this answers is not in the document]
Comment (punamdahiya) [reply package.json]: Done

**Comment by punamdahiya** [comment]
Tested and variationOverridePreference set in study.js isn't functional. PR updated to remove variationOverridePreference

File: README.md [hunk 12]
```diff
@@ -49,8 +43,6 @@ In this shield study, we have two experimental branches based on notification in
 2. notification-interval-longer - Cohort that’s shown notification in download panel after 2 days interval (since last notification shown and dismissed by clicking button Not Now)
 3. notification-interval-short - Cohort that’s shown notification after shorter interval - 1 day
 
-To force override a variation, set preference 'cloud.services.shield.variation' with the variation name
```
Comment (motin) [inline README.md]: You may re-add this, but phrased as "To force override a variation, clone the repo and set `testing.variationName` in `studySetup.js` to the variation name"

File: webextension/study.js [hunk 13]
```diff
@@ -18,7 +18,7 @@
  *
```
Comment (motin) [inline webextension/study.js]: This file should be named studySetup.js to match the small-study example

**Comment by motin** [comment]
> @motin Unsigned add-on should work on both beta and release. Can you please check the version,
WEE works on Firefox 60 and above.

You are right about the version 59 for release. I was probably just confused about the fact that it even installed. Add the version requirement to the manifest as per the small-study example:

```
  "applications": {
    "gecko": {
      "id": "shield-utils-test-addon@shield.mozilla.org",
      "strict_min_version": "60.0"
    }
  },
```

However, when running on beta (61.0b10), I get `TypeError: browser.study is undefined`.

> The easiest way to find the branch is look at pref cloud.services.interval.prompt, for Control branch it's value will be 0. An alternate way to is to look at about:telemetry. When add-on is installed, we are logging message 'addon_init' that should show branch.

Indeed I am placed in the control branch. Setting testing.variationName will put me in the control branch. Not that the current study.js boilerplate has testing.variation = null. It should be variationName with the latest utils v5 rc.

**Comment by punamdahiya** [comment]
As debugged today, browser.study error is seen with v5 in beta only and works for >=60  release, nightly and developer edition. @motin to look into shieldutils v5 fixes for
a) browser.study undefined error in beta b) handling testing.variationName code

**Comment by motin** [comment]
Indeed. Created an issues for the error on Beta: https://github.com/mozilla/shield-studies-addon-utils/issues/200 & regarding variationName, the issue has already been filed (https://github.com/mozilla/shield-studies-addon-utils/issues/195) but no fix has been submitted yet

**Comment by motin** [comment]
Good news - v5 is now official, so try installing it via `npm install shield-studies-addon-utils`, and if that does not work (maybe it is not published yet), try `npm install mozilla/shield-studies-addon-utils` to install it from Github.

**Comment by punamdahiya** [comment]
Updated to use mozilla/shield-studies-addon-utils master

**Comment by punamdahiya** [comment]
@motin @aswan PR is updated to point to v5 shield-utils RC. Please review if the integration looks good so we can merge this PR to master and have QA start testing it. Thanks!

**Comment by punamdahiya** [comment]
@motin Thanks, my understanding browser.study undefined error was showing up in in beta, merging PR to master for QA to verify and start testing.

**Comment by punamdahiya** [comment]
Firefox beta browser.study is undefined Issue tracked here https://github.com/mozilla/shield-studies-addon-utils/issues/200
